### DISCUSSION TITLE
Twig tag for composable wrapper component with children

### DISCUSSION THREAD

**Discussion by acalvino4** [craftcms/cms#12671] [Ideas]
Twig makes templates composable via `extends`, `include`, `embed`, `macro`, etc. but each of these seem to leave something behind in the way of composable modularity. [This article](https://dev.to/jakedohm_34/emulating-components-in-twig-cjo) explains most of what I'm looking for in the ideal construct, including the idea that components by default should only have access to variables that you explicitly pass (which is not true for `embed`). I agree with that article's author that embeds are the closest thing to what I want, but the other thing he doesn't mention their lack of direct support for a "children" slot. Sure, it's possible with embed blocks, but if you are actually using components for everything the way we're used to in the js world, and as recommended by tailwind, all those `{% block 'myblock' %}` / `{% endblock %}` statements start to feel pretty heavy. e.g.:

```twig
{% embed "button.twig" with {size: 'lg'} %}
    {% block buttontext %}
        <svg ... > --> An icon you want in your button
        <span>Button Content</span>
    {% endblock %}
{% endembed %}
```

I think the ideal api would be something like 

```twig
{% wrapper "button.twig" with {size: 'lg'} %}
    <svg ... > --> An icon you want in your button
    <span>Button Content</span>
{% endwrap %}
```

Basically it just loads the wrapped contents as a `children` variable that can be used within the wrapper definition. Also, only loads variables you pass to follow proper FP.

**Comment by brandonkelly** [comment]
How would the `{% wrapper %}` tag know where to inject the inner HTML, if the embedded template defines multiple HTML tags? E.g. `<div></div><div></div>`

Seems like it would be better for that template to explicitly define a `{{ block('children') }}` block somewhere, if it expects the parent template to be able to define nested HTML.

**Comment by acalvino4** [comment]
My general thought is that this construct would work exactly like `children` does in React - whatever content is between the opening and closing tags gets injected into your component as the `children` variable; then, in your component implementation, you just write `{children}` wherever you want that to go. In twig it would just be `{{ children }}`, which could be passed as a variable into the `renderTemplate` call of the implementation.

Now in React, you can assign a bunch of JSX to a variable and explicitly pass it to your component as a parameter, just as you can in twig, but it's not as immediately evident what's going on and doesn't make as good DX; hence react implemented the [children construct](https://beta.reactjs.org/learn/passing-props-to-a-component#passing-jsx-as-children).
Embeds do essentially do this in a more powerful way by allowing multiple slots - but it makes the base case more heavyweight.

**Reply by brandonkelly** [reply]
Gotcha. Looks like twigphp/Twig#3253 is basically what you’re after. Agree it would be a nice improvement – maybe voice your support over there.
